library/alloc/src/boxed.rs

```diff
@@ -518,20 +751,47 @@ impl<T: ?Sized> Box<T> {
     /// ```
     #[stable(feature = "box_leak", since = "1.26.0")]
     #[inline]
-    pub fn leak<'a>(b: Box<T>) -> &'a mut T
+    pub fn leak<'a>(b: Self) -> &'a mut T
     where
-        T: 'a, // Technically not needed, but kept to be explicit.
+        A: 'a,
     {
         unsafe { &mut *mem::ManuallyDrop::new(b).0.as_ptr() }
     }
 
+    /// Consumes and leaks the `Box`, returning a mutable reference,
+    /// `&'a mut T`, and the allocator. Note that the type `T` must outlive the chosen lifetime
+    /// `'a`. If the type has only static references, or none at all, then this
+    /// may be chosen to be `'static`.
+    ///
+    /// This function is mainly useful for data that lives for the remainder of
+    /// the program's life. Dropping the returned reference will cause a memory
+    /// leak. If this is not acceptable, the reference should first be wrapped
+    /// with the [`Box::from_raw`] function producing a `Box`. This `Box` can
+    /// then be dropped which will properly destroy `T` and release the
+    /// allocated memory.
+    ///
+    /// Note: this is an associated function, which means that you have
+    /// to call it as `Box::leak(b)` instead of `b.leak()`. This
+    /// is so that there is no conflict with a method on the inner type.
+    #[unstable(feature = "allocator_api", issue = "32838")]
+    #[inline]
+    pub fn leak_with_alloc<'a>(b: Self) -> (&'a mut T, A)
+    where
+        A: 'a,
+    {
+        let manually_drop = mem::ManuallyDrop::new(b);
+        let alloc = unsafe { ptr::read(&manually_drop.1) };
```

Comment:
You probably want `MaybeUninit` as you're effectively un-initializing part of the data contained within `ManuallyDrop` which is somewhat sketchy as the value is later used again to obtain `reference`.

Reply:
Ah, nice catch. So I think this way this should be safe?
```rust
let mut boxed = mem::MaybeUninit::new(b);
unsafe {
    let reference = &mut *(*b.as_mut_ptr()).0.as_ptr();
    let alloc = ptr::read(&(*b.as_mut_ptr()).1);
    (reference, alloc)
}
```

Or would it also be safe to obtain the reference **before** reading the allocator?